precompile/modules/initia_stdlib/sources/dex.move

```diff
@@ -981,14 +1008,95 @@ module initia_std::dex {
+    ///
+    /// https://move-book.com/programmability/hot-potato-pattern.html
+    public fun flash_swap(
+        pair: Object<Config>,
+        offer_coin: Object<Metadata>,
+        offer_amount: u64,
+        min_return: Option<u64>
+    ): (FungibleAsset, FlashSwapLoan) acquires Config, Pool, FlashSwap {
+        // zero offer amount would be invalidated in swap_simulation
+        let return_amount = get_swap_simulation(pair, offer_coin, offer_amount);
+        assert!(
+            option::is_none(&min_return)
+                || *option::borrow(&min_return) <= return_amount,
+            error::invalid_state(EMIN_RETURN)
+        );
+
+        let pair_addr = object::object_address(&pair);
+        let pool = borrow_global<Pool>(pair_addr);
+        let pool_config = borrow_global<Config>(pair_addr);
+        let pair_signer = &object::generate_signer_for_extending(&pool_config.extend_ref);
+        let pair_addr = signer::address_of(pair_signer);
+
+        // check recursive flash swap
+        assert!(
+            !exists<FlashSwap>(pair_addr),
+            error::invalid_state(ERECURSIVE_FLASH_SWAP)
+        );
+
+        let coin_a_metadata = fungible_asset::store_metadata(pool.coin_a_store);
+
+        // withdraw the return coin
+        let (return_coin, coin_a_borrow_amount, coin_b_borrow_amount) =
+            if (object::object_address(&offer_coin)
+                == object::object_address(&coin_a_metadata)) {
+                (
+                    fungible_asset::withdraw(
+                        pair_signer, pool.coin_b_store, return_amount
+                    ),
+                    offer_amount,
+                    0
+                )
+            } else {
+                (
+                    fungible_asset::withdraw(
+                        pair_signer, pool.coin_a_store, return_amount
+                    ),
+                    0,
+                    offer_amount
+                )
+            };
+
+        // store flash swap to prevent recursive flash swap
+        move_to(pair_signer, FlashSwap { coin_a_borrow_amount, coin_b_borrow_amount });
+
+        (return_coin, FlashSwapLoan { pair_addr })
+    }
+
+    public fun repay_flash_swap(
+        repay_fa: FungibleAsset, loan: FlashSwapLoan
+    ) acquires Pool, FlashSwap {
+        let FlashSwapLoan { pair_addr } = loan;
+        let FlashSwap { coin_a_borrow_amount, coin_b_borrow_amount } =
+            move_from(pair_addr);
+
+        let pool = borrow_global_mut<Pool>(pair_addr);
+        let (repay_amount, repay_store) =
+            if (coin_a_borrow_amount > 0) {
+                (coin_a_borrow_amount, pool.coin_a_store)
+            } else {
+                (coin_b_borrow_amount, pool.coin_b_store)
+            };
+
+        assert!(
+            fungible_asset::amount(&repay_fa) == repay_amount,
+            error::invalid_argument(EFAILED_TO_REPAY_FLASH_SWAP)
+        );
+        fungible_asset::deposit(repay_store, repay_fa);
```

Comment:
does initia support [dispatchable FA](https://aptos.dev/en/build/smart-contracts/fungible-asset#dispatchable-fungible-asset-advanced)?

if so, we should try to avoid `fungible_asset::deposit/withdraw`, see https://aptos.dev/en/build/smart-contracts/fungible-asset#limitations

instead, rely on `primary_fungible_store` api wherever possible

Reply:
in current fungible asset implementation, if we call `fungible_asset::deposit or withdraw` directly it may abort if the `FungibleAsset` has dispatch function with `abort_on_dispatch = true` [here](https://github.com/initia-labs/movevm/blob/e4925477a550743ee450d340a590549ae82fb2b4/precompile/modules/initia_stdlib/sources/fa/fungible_asset.move#L725) and [here](https://github.com/initia-labs/movevm/blob/e4925477a550743ee450d340a590549ae82fb2b4/precompile/modules/initia_stdlib/sources/fa/fungible_asset.move#L737).

I think it is reasonable to avoid `DispatchableFA` in dex for now to prevent unknown problems.